pilot/pkg/networking/util/util.go

```diff
@@ -256,22 +261,89 @@ func ConvertLocality(locality string) *core.Locality {
 		return nil
 	}
 
+	region, zone, subzone := SplitLocality(locality)
+	return &core.Locality{
+		Region:  region,
+		Zone:    zone,
+		SubZone: subzone,
+	}
+}
+
+func LocalityMatch(proxyLocality *core.Locality, ruleLocality string) bool {
+	ruleRegion, ruleZone, ruleSubzone := SplitLocality(ruleLocality)
+	regionMatch := ruleRegion == "*" || proxyLocality.GetRegion() == ruleRegion
+	zoneMatch := ruleZone == "*" || ruleZone == "" || proxyLocality.GetZone() == ruleZone
+	subzoneMatch := ruleSubzone == "*" || ruleSubzone == "" || proxyLocality.GetSubZone() == ruleSubzone
+
+	if regionMatch && zoneMatch && subzoneMatch {
+		return true
+	}
+	return false
+}
+
+func SplitLocality(locality string) (region, zone, subzone string) {
 	items := strings.Split(locality, "/")
 	switch len(items) {
 	case 1:
-		return &core.Locality{
-			Region: items[0],
-		}
+		return items[0], "", ""
 	case 2:
-		return &core.Locality{
-			Region: items[0],
-			Zone:   items[1],
-		}
+		return items[0], items[1], ""
 	default:
-		return &core.Locality{
-			Region:  items[0],
-			Zone:    items[1],
-			SubZone: items[2],
+		return items[0], items[1], items[2]
+	}
+}
+
+func LbPriority(proxyLocality, endpointsLocality *core.Locality) int {
+	if proxyLocality.GetRegion() == endpointsLocality.GetRegion() {
+		if proxyLocality.GetZone() == endpointsLocality.GetZone() {
+			if proxyLocality.GetSubZone() == endpointsLocality.GetSubZone() {
+				return 0
+			}
+			return 1
 		}
+		return 2
+	}
+	return 3
+}
+
+// return a shallow copy cluster
+func CloneCluster(cluster *xdsapi.Cluster) xdsapi.Cluster {
```

Comment:
Can you explain in the comments why do we need all the cloning ? Can it be avoided ? Any perf-sensitive  code should have some explanation.

Reply:
If we knew the list of all localities for every proxy that could connect to Pilot, in advance, we could have created caches per locality when creating the pushContext for this epoch. Knowing all the localities in advance would require scanning through all the connected proxies in ADS, collecting the localities and then creating the caches based on the localities of the connected proxies. But then, the proxies connected to ADS server keep changing. proxies from new localities may connect later, while all proxies from one locality may go away.

sumOfAllServiceInstances != numberOfProxiesConnectedToSinglePilot. So we cannot derive this info from service instances as well. We could do something optimistically, but we would still have to fallback to creating the clusters from scratch if a proxy from a locality that was previously unknown connects to Pilot. 
Rather than building the clusters from scratch and spending more processing time, the shallow cloning is very light weight. It reuses most of the pointer fields inside the cluster (tls contexts, etc.), and updates only the locality settings.